### PR TITLE
(Issue 209, 275) Make library truly header-only

### PR DESCRIPTION
**Pull request by nrkramer**
Picking up from https://github.com/ZigRazor/CXXGraph/pull/224

*A good pull request has the following information:*
- A reference to a related issue in your repository.
  - https://github.com/ZigRazor/CXXGraph/issues/209
  - https://github.com/ZigRazor/CXXGraph/issues/275
- A description of the changes proposed in the pull request.
  1. Remove dependency on generated files by checking-in `include/CXXGraphConfig.h`. This means any changes to version will change this file, however it makes the library truly header-only.
  2. Refractor how dependencies are fetched and included in the `CMakeLists.txt` for tests and benchmarking.
  3. Streamline how files are written/read in `include/Graph/Graph.hpp`. This sets us up for removing the dependency on compression in the distributable library.
  4. Remove compression from the header using a compiler define. Is this the cleanest way? Probably not. Could definitely be more clean by moving optional functions to a separate header. But this works for now.
- @mentions of the person or team responsible for reviewing proposed changes. ( optional ) 
  - @ZigRazor 
  - @amerinor01 (optional)

### CODE REVIEW

**Comment by nrkramer** [comment]
@ZigRazor I was unable to run tests on my setup because I don't use Mingw on my Windows machine.

Would you mind running them for me? Feel free to push any fixes directly to this PR

File: .gitignore [hunk 1]
```diff
@@ -35,7 +35,5 @@
 build/
 # ignore the packaging folder contents from git
 packaging/
-# graph config is automatically setup, so we can ignore this
-include/CXXGraphConfig.h
```
Comment (nrkramer) [inline .gitignore:39]: Removed so that we don't depend on generated files during configure time.

File: benchmark/CMakeLists.txt [hunk 2]
```diff
@@ -9,17 +9,47 @@ if(BENCHMARK)
...
+    GIT_TAG origin/master
+    OPTIONS "WITH_APPS OFF"
+    )
+
+    FetchContent_Declare(
+    zlib
+    GIT_REPOSITORY https://github.com/madler/zlib.git
+    GIT_TAG v1.2.13
+    OPTIONS "CMAKE_POSITION_INDEPENDENT_CODE True"
+    )
```
Comment (nrkramer) [inline benchmark/CMakeLists.txt:24]: Fetch both openssl and zlib.

openssl will be removed soon: https://github.com/ZigRazor/CXXGraph/issues/274

File: include/CXXGraphConfig.h [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// the configured options and settings for CXXGraph
+#define CXXGraph_VERSION_MAJOR 1
+#define CXXGraph_VERSION_MINOR 0
+#define CXXGraph_VERSION_PATCH 0
```
Comment (nrkramer) [inline include/CXXGraphConfig.h:4]: We're checking in this file now.

Any change to the project version will be reflected here.

File: include/Graph/Graph.hpp [hunk 4]
```diff
@@ -85,24 +88,38 @@ class Graph {
...
-                           InputOutputFormat format);
-  void recreateGraphFromReadFiles(
+  void writeGraphToStream(const std::ostream &oGraph,
+                          const std::ostream &oNodeFeat,
+                          const std::ostream &oEdgeWeight, const char &sep,
+                          bool writeNodeFeat, bool writeEdgeWeight) const;
+  void readGraphFromStream(const std::istream &iGraph,
+                           const std::istream &iNodeFeat,
+                           const std::istream &iEdgeWeight, bool readNodeFeat,
+                           bool readEdgeWeight);
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:98]: R/W with generic streams is much more portable in application (for example, use in network streams, binary streams, etc etc)

File: include/Graph/Graph.hpp [hunk 5]
```diff
@@ -85,24 +88,38 @@ class Graph {
...
+  virtual int writeToFile(
+      InputOutputFormat format = InputOutputFormat::STANDARD_CSV,
+      const std::string &workingDir = ".",
+      const std::string &OFileName = "graph", bool compress = false,
+      bool writeNodeFeat = false, bool writeEdgeWeight = false) const;
+  virtual int readFromFile(
+      InputOutputFormat format = InputOutputFormat::STANDARD_CSV,
+      const std::string &workingDir = ".",
+      const std::string &OFileName = "graph", bool compress = false,
+      bool readNodeFeat = false, bool readEdgeWeight = false);
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:115]: Make the version that uses compression private API

File: include/Graph/Graph.hpp [hunk 6]
```diff
@@ -503,17 +520,15 @@ class Graph {
...
    * @param writeEdgeWeight Indicates if export also Edge Weights
    * @return 0 if all OK, else return a negative value
    */
   virtual int writeToFile(
       InputOutputFormat format = InputOutputFormat::STANDARD_CSV,
       const std::string &workingDir = ".",
-      const std::string &OFileName = "graph", bool compress = false,
-      bool writeNodeFeat = false, bool writeEdgeWeight = false) const;
+      const std::string &OFileName = "graph", bool writeNodeFeat = false,
+      bool writeEdgeWeight = false) const;
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:531]: ⚠️ Public API change. We're removing the ability to compress. Users can use their own compression wrappers if they want.

File: include/Graph/Graph.hpp [hunk 7]
```diff
@@ -524,16 +539,15 @@ class Graph {
...
    * @param readEdgeWeight Indicates if import also Edge Weights
    * @return 0 if all OK, else return a negative value
    */
   virtual int readFromFile(
       InputOutputFormat format = InputOutputFormat::STANDARD_CSV,
       const std::string &workingDir = ".",
-      const std::string &OFileName = "graph", bool compress = false,
-      bool readNodeFeat = false, bool readEdgeWeight = false);
+      const std::string &OFileName = "graph", bool readNodeFeat = false,
+      bool readEdgeWeight = false);
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:550]: ⚠️ Public API change. We're removing the ability to compress. Users can use their own compression wrappers if they want.

File: test/KosarajuTest.cpp [hunk 8]
```diff
@@ -1,5 +1,3 @@
-#include <zlib.h>
-
```
Comment (nrkramer) [inline test/KosarajuTest.cpp:2]: Removed this hanging include

**Comment by nrkramer** [comment]
Would also be nice to do a version bump to 1.0.1 after this is merged

File: .gitignore (hunk 1 above)
Comment (ZigRazor) [reply .gitignore:39]: OK

File: test/KosarajuTest.cpp (hunk 8 above)
Comment (ZigRazor) [reply test/KosarajuTest.cpp:2]: OK!

**Comment by ZigRazor** [comment]
Some checks does not succed but is another problem on the repo, I will release a new version with this pull request

**Pull request merged**